File: brainplus-engine/src/api/anthropic.rs
//! Anthropic Messages API 流式客户端
//! 对齐 src/lib/api/anthropic.ts

use super::{ChatMessage, ProviderConfig, StreamEvent, ToolDef};
use futures::StreamExt;
use serde_json::Value;

pub async fn stream_anthropic(
    config: &ProviderConfig,
    messages: &[ChatMessage],
    tools: &[ToolDef],
    system_prompt: Option<&str>,
    max_tokens: Option<u32>,
    mut tx: tokio::sync::mpsc::Sender<StreamEvent>,
) -> Result<Usage, String> {
    let base_url = config.base_url.as_deref().unwrap_or("https://api.anthropic.com");
    let url = format!("{}/v1/messages", base_url.trim_end_matches('/'));

    // 构建 anthropic 格式的 tool 定义
    let anthropic_tools: Vec<Value> = tools.iter().map(|t| {
        serde_json::json!({
            "name": t.name,
            "description": t.description,
            "input_schema": t.input_schema,
        })
    }).collect();

    // 分离 system 消息
    let system = system_prompt.unwrap_or("You are a coding assistant.").to_string();

    // 构建 messages（Anthropic 格式不需要 system role）
    let anthropic_msgs: Vec<Value> = messages.iter().map(|m| {
        let mut obj = serde_json::json!({"role": m.role, "content": m.content});
        if let Some(ref tc) = m.tool_calls {
            let tool_blocks: Vec<Value> = tc.iter().map(|c| {
                serde_json::json!({
                    "type": "tool_use",
                    "id": c.id,
                    "name": c.name,
                    "input": c.input,
                })
            }).collect();
            obj["content"] = serde_json::json!(tool_blocks);
        }
        if let Some(ref tci) = m.tool_call_id {
            obj["content"] = serde_json::json!([{
                "type": "tool_result",
                "tool_use_id": tci,
                "content": m.content,
            }]);
            obj["role"] = serde_json::Value::String("user".into());
        }
        obj
    }).collect();

    let mut body = serde_json::json!({
        "model": config.model_id,
        "messages": anthropic_msgs,
        "system": system,
        "max_tokens": max_tokens.unwrap_or(4096),
        "stream": true,
    });
    if !anthropic_tools.is_empty() {
        body["tools"] = serde_json::json!(anthropic_tools);
    }

    let client = reqwest::Client::new();
    let response = client.post(&url)
        .header("x-api-key", &config.api_key)
        .header("anthropic-version", "2023-06-01")
        .header("Content-Type", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("Anthropic API 请求失败: {e}"))?;

    let status = response.status();
    if !status.is_success() {
        let err_body = response.text().await.unwrap_or_default();
        return Err(format!("Anthropic HTTP {}: {}", status.as_u16(), &err_body[..300.min(err_body.len())]));
    }

    let mut stream = response.bytes_stream();
    let mut buffer = String::new();
    let mut input_tokens = 0u32;
    let mut output_tokens = 0u32;
    let mut finish_reason = "stop".to_string();
    let mut pending_tool_call: Option<(String, String)> = None; // (id, name)
    let mut pending_input = String::new();

    while let Some(chunk) = stream.next().await {
        let chunk = chunk.map_err(|e| format!("SSE 读取错误: {e}"))?;
        buffer.push_str(&String::from_utf8_lossy(&chunk));

        while let Some(pos) = buffer.find("\n\n") {
            let event_str = buffer[..pos].to_string();
            buffer = buffer[pos + 2..].to_string();

            for line in event_str.lines() {
                let line = line.trim();
                if line.is_empty() || !line.starts_with("data: ") {
                    continue;
                }

                let data = &line[6..];
                if data == "[DONE]" { continue; }

                let event: Value = match serde_json::from_str(data) {
                    Ok(v) => v,
                    Err(_) => continue,
                };

                match event["type"].as_str() {
                    Some("message_start") => {
                        if let Some(usage) = event["message"].get("usage") {
                            input_tokens = usage["input_tokens"].as_u64().unwrap_or(0) as u32;
                        }
                    }

                    Some("content_block_start") => {
                        let block = &event["content_block"];
                        if block["type"].as_str() == Some("tool_use") {
                            let id = block["id"].as_str().unwrap_or("").to_string();
                            let name = block["name"].as_str().unwrap_or("").to_string();
                            pending_tool_call = Some((id, name));
                            pending_input.clear();
                        }
                    }

                    Some("content_block_delta") => {
                        let delta = &event["delta"];
                        match delta["type"].as_str() {
                            Some("text_delta") => {
                                let text = delta["text"].as_str().unwrap_or("");
                                let _ = tx.send(StreamEvent::TextDelta { text: text.to_string() }).await;
                            }
                            Some("thinking_delta") => {
                                let text = delta["thinking"].as_str().unwrap_or("");
                                let _ = tx.send(StreamEvent::ReasoningDelta { text: text.to_string() }).await;
                            }
                            Some("input_json_delta") => {
                                pending_input.push_str(delta["partial_json"].as_str().unwrap_or(""));
                            }
                            _ => {}
                        }
                    }

                    Some("content_block_stop") => {
                        if let Some((id, name)) = pending_tool_call.take() {
                            // 尝试解析累积的 JSON 输入
                            let input: Value = if pending_input.is_empty() {
                                serde_json::json!({})
                            } else {
                                serde_json::from_str(&pending_input).unwrap_or(serde_json::json!({}))
                            };
                            let _ = tx.send(StreamEvent::ToolCallStart { id, name, input }).await;
                        }
                    }

                    Some("message_delta") => {
                        if let Some(delta) = event.get("delta") {
                            finish_reason = delta["stop_reason"].as_str().unwrap_or("stop").to_string();
                        }
                        if let Some(usage) = event.get("usage") {
                            output_tokens = usage["output_tokens"].as_u64().unwrap_or(0) as u32;
                        }
                    }

                    Some("message_stop") => {
                        // Final event
                    }

                    _ => {}
                }
            }
        }
    }

    let _ = tx.send(StreamEvent::Done {
        finish_reason: finish_reason.clone(),
        input_tokens,
        output_tokens,
    }).await;

    Ok(Usage { input_tokens, output_tokens })
}

// re-export types via super
use super::Usage;

File: src/lib/api/anthropic.ts
/**
 * Anthropic API provider — 使用 @anthropic-ai/sdk 直调
 * 对齐 CC 的 query.ts → claude.ts 路径
 */
import Anthropic from '@anthropic-ai/sdk'
import type { ToolDef, StreamEvent, ChatMessage, ProviderConfig, Usage } from './types'

// Anthropic SDK 事件类型
type RawEvent = {
  type: string
  index?: number
  content_block?: { type: string; id?: string; name?: string; text?: string; thinking?: string }
  delta?: { type: string; text?: string; partial_json?: string; thinking?: string }
  message?: { id: string; model: string; usage: { input_tokens: number } }
  usage?: { output_tokens: number }
}

function createClient(config: ProviderConfig): Anthropic {
  return new Anthropic({
    apiKey: config.apiKey,
    baseURL: config.baseUrl || undefined,
    // CC 同款 beta headers
    defaultHeaders: {
      'anthropic-beta':
        'interleaved-thinking-2025-05-14,context-management-2025-06-27,prompt-caching-scope-2026-01-05,mid-conversation-system-2026-04-07',
    },
  })
}

/**
 * 把 stardust 工具定义转成 Anthropic API 工具格式
 */
function convertTools(tools: Record<string, ToolDef>): Anthropic.Tool[] {
  return Object.entries(tools).map(([name, tool]) => ({
    name,
    description: tool.description,
    input_schema: (tool.inputSchema || { type: 'object', properties: {} }) as Record<string, unknown>,
  })) as unknown as Anthropic.Tool[]
}

/**
 * 把 stardust 消息转成 Anthropic API 消息格式
 */
function convertMessages(
  messages: ChatMessage[],
  system?: string,
): { system?: string | Array<{ type: 'text'; text: string }>; messages: Anthropic.MessageParam[] } {
  const msgs: Anthropic.MessageParam[] = []

  for (const m of messages) {
    if (m.role === 'system') {
      system = system ? system + '\n\n' + m.content : m.content
      continue
    }
    if (m.role === 'user') {
      msgs.push({ role: 'user', content: m.content })
    } else if (m.role === 'assistant') {
      // 构建 assistant 内容：文本 + 工具调用
      const content: Anthropic.ContentBlockParam[] = []
      if (m.content) {
        content.push({ type: 'text', text: m.content })
      }
      if (m.toolCalls && m.toolCalls.length > 0) {
        for (const tc of m.toolCalls) {
          content.push({
            type: 'tool_use',
            id: tc.id,
            name: tc.name,
            input: tc.input as Record<string, unknown>,
          })
        }
      }
      msgs.push({
        role: 'assistant',
        content: content.length > 0 ? content : m.content || '',
      })
    } else if (m.role === 'tool') {
      // 工具结果转换为 Anthropic 格式：user 消息含 tool_result block
      msgs.push({
        role: 'user',
        content: [{
          type: 'tool_result',
          tool_use_id: m.toolCallId || '',
          content: m.content,
        }],
      })
    }
  }

  return { system: system || undefined, messages: msgs }
}

/**
 * 流式调用 Anthropic API，yield 统一的 StreamEvent
 */
export async function* streamAnthropic(
  config: ProviderConfig,
  messages: ChatMessage[],
  tools: Record<string, ToolDef>,
  systemPrompt?: string,
  signal?: AbortSignal,
  maxOutputTokens?: number,
  thinkingBudgetTokens?: number,
): AsyncGenerator<StreamEvent> {
  const client = createClient(config)
  const { system, messages: apiMessages } = convertMessages(messages, systemPrompt)
  const apiTools = convertTools(tools)

  const maxTokens = maxOutputTokens ?? 8192
  const streamOptions: Record<string, unknown> = {
    model: config.modelId,
    max_tokens: maxTokens,
    temperature: 0.3,
    system: system || undefined,
    messages: apiMessages,
    tools: apiTools.length > 0 ? apiTools : undefined,
  }

  // 思考预算：需要 max_tokens > budget_tokens
  if (thinkingBudgetTokens && thinkingBudgetTokens > 0) {
    streamOptions.thinking = {
      type: 'enabled',
      budget_tokens: Math.min(thinkingBudgetTokens, maxTokens - 1024),
    }
  }

  const stream = client.beta.messages.stream(streamOptions as any)

  let toolUseId = ''
  let toolUseName = ''
  let toolInputAccum = ''
  let currentBlockType = ''
  let inputTokens = 0
  let outputTokens = 0

  try {
    for await (const event of stream) {
      if (signal?.aborted) break

      const evt = event as unknown as RawEvent

      switch (evt.type) {
        case 'message_start':
          inputTokens = evt.message?.usage?.input_tokens ?? 0
          break

        case 'content_block_start': {
          const block = evt.content_block
          if (!block) break
          currentBlockType = block.type
          if (block.type === 'tool_use') {
            toolUseId = block.id || ''
            toolUseName = block.name || ''
            toolInputAccum = ''
          } else if (block.type === 'thinking' && block.thinking) {
            yield {
              type: 'reasoning-delta',
              text: block.thinking,
              reasoningId: `reasoning-${evt.index ?? 0}`,
            }
          }
          break
        }

        case 'content_block_delta': {
          const delta = evt.delta
          if (!delta) break
          if (delta.type === 'text_delta' && delta.text) {
            yield { type: 'text-delta', text: delta.text }
          } else if (delta.type === 'input_json_delta' && delta.partial_json) {
            toolInputAccum += delta.partial_json
          } else if (delta.type === 'thinking_delta' && delta.thinking) {
            yield {
              type: 'reasoning-delta',
              text: delta.thinking,
              reasoningId: `reasoning-${evt.index ?? 0}`,
            }
          }
          break
        }

        case 'content_block_stop':
          if (currentBlockType === 'tool_use' && toolUseId && toolUseName) {
            yield {
              type: 'tool-call-start',
              id: toolUseId,
              name: toolUseName,
              input: toolInputAccum,
            }
            toolUseId = ''
            toolUseName = ''
            toolInputAccum = ''
          }
          currentBlockType = ''
          break
      }
    }
  } catch (e: any) {
    if (e.name === 'AbortError' || signal?.aborted) {
      yield { type: 'done', finishReason: 'abort', usage: { inputTokens, outputTokens } }
      return
    }
    throw e
  }

  // 获取最终 usage
  try {
    const final = await stream.finalMessage()
    outputTokens = final.usage.output_tokens
  } catch {}

  const usage: Usage = { inputTokens, outputTokens }
  yield { type: 'done', finishReason: 'stop', usage }
}
